### ISSUE TITLE
Generic.Files.LineLength should count line width instead of char counts

### ISSUE THREAD

**Issue by pokehanai** [squizlabs/PHP_CodeSniffer#534]
Currently `phpcs --encoding=utf8` checks line length by number of characters.
But while the rule is for readability for human, it should care of display width of each line.

PHP has [mb_strwidth()][] for this situation.
[mb_strwidth()]: http://php.net/mb_strwidth

| line   | current | expected |
| :---   |     --: |      --: |
| aaa    |       3 |        3 |
| ☆☆☆ |       3 |        6 |

**Comment by aik099** [comment]
I wonder how you managed to type these symbols into PHP code.

Is there a similar function for iconv, because we're using iconv.

**Comment by pokehanai** [comment]
hmm, iconv seems not to have that functionality.

Speaking of typing symbols, asians and others use [IME][], so it's easy with it :)
And I as a Japanese basically write string literals in Japanese language for end users and code comments for my teams. So to check lines in width is really needed. 
[IME]: http://en.wikipedia.org/wiki/Input_method

**Comment by gsherwood** [comment]
Sorry, but this isn't something that I can do.

Letters and symbols do not have sizes; they are only character codes. The font you choose gives them their size (specifically their width, in your case). There is no way to look at a character and determine the width.

If you have a very specific use case, you can always write a custom sniff that takes in account the specific characters you use and the font and font size you typically use, to estimate line width. But it's not something I would ever consider doing in the Generic line length sniff.

**Comment by pokehanai** [comment]
The Generic routine don't need to calculate width in pixels. But it need to do at least what general text editors do.

c.f. http://en.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms

> Sorry, but this isn't something that I can do.

You can do it with [mb_strwidth()][], which is as same as what text editors do to count columns, and the meaning of resulting number is same with what `tab width` do.

...ah, mb_strwidth() looks imperfect for CJK.
There are some punctuations and symbols that have different width with of text editor.

I found a library which deals it well: https://github.com/zonuexe/php-stringwidth

Code to calculate each token would be something like below
```php
function calculateTokenWidth($token, $encoding, $modeCJK = false) {
    $utf8Token = mb_convert_encoding($token, 'UTF-8', $encoding);
    return $modeCJK ? \Teto\StringWidth::asDouble($utf8Token) :
                      \Teto\StringWidth::asSingle($utf8Token);
}
```

And with white space character calculation logic, we can finally get a fair line width that Generic should deal with.

[mb_strwidth()]: http://php.net/manual/en/function.mb-strwidth.php

**Comment by gsherwood** [comment]
Thanks for all that information, but I'm not changing the way the Generic sniff counts line length. As I said, you can always write a custom sniff for your own needs.
